dagshub/data_engine/model/datasource.py

```diff
@@ -543,49 +617,111 @@ def _upload_metadata(self, metadata_entries: List[DatapointMetadataUpdateEntry])
+        mlflow.log_dict(self._to_dict(), artifact_name)
+        log_message(f'Saved the datasource state to MLflow (run "{run.info.run_name}") as "{artifact_name}"')
+
     def save_to_file(self, path: str = ".", name: str = "") -> str:
         """
-        [EXPERIMENTAL]
-        Saves a text file representing the datasource or dataset information to a file which can be committed to Git.
+        Saves a text file representing the current state of datasource or dataset
+        to a file ``<name>.json`` which can be committed to Git.
         Useful for connecting code versions to the datasource used for training.
 
         .. note::
-            Does not save dataset contents, only a pointer file.
+            Does not save dataset contents.
 
         Args:
-            path: Path to save the datasource or dataset file in, defaults to current directory
-            name: Optional: name of the file. if not provided, datasource or dataset name will be used instead.
+            path: Path to the directory where to save the datasource or dataset file.
+                Defaults to current working directory.
+            name: Optional: name of the file. if not provided, datasource name will be used instead.
 
         Returns:
             The path to the saved file
         """
-        res = DatasourceFileStruct(
-            id=self._source.id,
-            repo=self._source.repo,
-            name=self._source.name,
-            datasetQuery=self.serialize_gql_query_input()
-        )
-        file_path = os.path.join(path, (name or res.name) + ".dagshub")
+        file_path = os.path.join(path, (name or self.source.name) + ".json")
+        res = self._to_dict()
         with open(file_path, "w") as file:
-            file.write(json.dumps(res.to_dict(), indent=4))
+            file.write(json.dumps(res, indent=4))
         log_message(f"Datasource saved to '{file_path}'")
 
         return file_path
 
+    def _serialize(self) -> DatasourceSerializedState:
+        res = DatasourceSerializedState(
+            repo=self.source.repo,
+            datasource_id=self.source.id,
+            datasource_name=self.source.name,
+            query=self.serialize_gql_query_input(),
+            timestamp=datetime.datetime.now().timestamp(),
+            modified=self.is_query_different_from_dataset,
+        )
+        if self.assigned_dataset is not None:
+            res.dataset_id = self.assigned_dataset.dataset_id
+            res.dataset_name = self.assigned_dataset.dataset_name
+        return res
+
+    def _to_dict(self) -> Dict:
+        res = self._serialize().to_dict()
+        # Skip Nones in the result
+        res = {k: v for k, v in res.items() if v is not None}
+        return res
+
+    @property
+    def is_query_different_from_dataset(self) -> Optional[bool]:
+        if self.assigned_dataset is None:
+            return None
+        return self._query.to_dict() != self.assigned_dataset.query.to_dict()
```

Comment:
Will this comparison actually work?
Seems like you need a deep comparison

Reply:
I checked and it actually works. A couple of the tests rely on that.
Works because you only need deep comparison if the subelements in the dict are objects. `query.to_dict()` returns dict with only primitive keys and values, so it's fine.

Reply:
I think you're right and it might not work with the nested query? I'll make sure to check, thank you.